core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java

```diff
@@ -168,39 +173,56 @@ public long getSizeInBytes()
         return sizeInBytes;
     }
 
-    private void copyBytes(Block block, int[] lengths, int[] positions, int count, int[] targetOffsets, int targetOffsetsIndex, int newByteCount)
+    private void copyBytes(Slice rawSlice, int[] lengths, int[] sourceOffsets, int count, int newByteCount)
     {
         ensureBytesCapacity(getCurrentOffset() + newByteCount);
 
-        for (int i = 0; i < count; i++) {
-            int position = positions[i];
-            if (!block.isNull(position)) {
+        Object base = rawSlice.getBase();
+        if (base instanceof byte[] && rawSlice.getAddress() == ARRAY_BYTE_BASE_OFFSET) {
```

Comment:
Since you are actively avoiding use of `Slice#getBytes(int index, byte[] destination, int destinationIndex, int length)` here, could you check if the benefit is coming from simply skipping the range checks done inside that method or is `System.arraycopy` faster than `unsafe.copyMemory` ?
If it is the latter, you could just improve `Slice#getBytes` implementation to use `System.arraycopy` for byte array case (it would also be interesting to know why `System.arraycopy` is faster than `unsafe.copyMemory`). If it is the former, you could look into providing an unchecked version of `Slice#getBytes` in Slice.

Reply:
I attempted to move `System.arraycopy` to `Slice.getBytes`, but that was slower than the current code in JMH benchmarks. So I suspect at least some of the improvement here is due to fewer positions checks and/or manual inlining. For the time being I've applied some fixups but kept same approach.